packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp

```diff
@@ -0,0 +1,306 @@
+    std::vector<int> owners_vec;
+    std::vector<gno_t> boundary_vec;
+    for(gno_t i = 0; i < num_boundary_edges*2; i++){
+      boundary_vec.push_back(boundary_edges[i]);
+      owners_vec.push_back(0);
+    }
+    Teuchos::ArrayView<gno_t> boundary = Teuchos::arrayViewFromVector(boundary_vec);
+    Teuchos::ArrayView<int> owners = Teuchos::arrayViewFromVector(owners_vec);
+  
+    //get the owning process for each vertex in the boundary list
+    mapWithCopies->getRemoteIndexList(boundary,owners);
+    //store the number of processes for future use
+    int nprocs = problemComm->getSize();
+  
+    //send boundary edges to any remote processes that own an endpoint
+    std::vector<int> sendcnts;
+    for(int i = 0; i < nprocs; i++) sendcnts.push_back(0);
+    for(int i = 0; i < num_boundary_edges*2; i+= 2){
+      if(owners[i] != -1){
+        sendcnts[owners[i]] +=2;
+      }
+      if(owners[i+1] != -1 && owners[i] != owners[i+1]){
+        sendcnts[owners[i+1]] +=2;
+      }
+    }
+    std::vector<int> recvcnts;
+    for(int i = 0; i < nprocs; i++) recvcnts.push_back(0);
+    //send the number of vertex IDs to be sent.
+    Teuchos::ArrayView<const int> sendCount = Teuchos::arrayViewFromVector(sendcnts);
+    Teuchos::ArrayView<int> recvCount = Teuchos::arrayViewFromVector(recvcnts);
+  
+    //create sdispls and sentcount to build sendbuf
+    std::vector<int> sdispls;
+    sdispls.push_back(0);
+    for(int i = 1; i < nprocs; i++){
+      sdispls.push_back(sdispls[i-1] + sendcnts[i-1]);
+    }
+
+    std::vector<int> sentcount;
+    for(int i = 0; i < nprocs; i++){
+      sentcount.push_back(0);
+    }
+    int sendsize = 0;
+    int recvsize = 0;
+    for(int i = 0; i < nprocs; i++){
+      sendsize += sendCount[i];
+      recvsize += recvCount[i];
+    }
+  
+    std::vector<gno_t> sendbuf(sendsize,0);
+    for(int i = 0; i > num_boundary_edges*2; i+=2){
+      if(owners[i] != me){
+        int proc_to_send = owners[i];
+        int sendbufidx = sdispls[proc_to_send] + sentcount[proc_to_send];
+        sentcount[proc_to_send] += 2;
+        sendbuf[sendbufidx++] = boundary[i];
+        sendbuf[sendbufidx++] = boundary[i+1];
+      }
+      if(owners[i+1] != me && owners[i] != owners[i+1]){
+        int proc_to_send = owners[i];
+        int sendbufidx = sdispls[proc_to_send] + sentcount[proc_to_send];
+        sentcount[proc_to_send] += 2;
+        sendbuf[sendbufidx++] = boundary[i];
+        sendbuf[sendbufidx++] = boundary[i+1];
+      }
+    }
+ 
+    //Do the final communication back to each remote process 
+    Teuchos::ArrayRCP<gno_t> recvbuf;
+    Teuchos::ArrayView<const gno_t> sendBuf = Teuchos::arrayViewFromVector(sendbuf);
+    Zoltan2::AlltoAllv(*problemComm,*env,sendBuf,sendCount,recvbuf,recvCount);
+
+    //hash for the boundary-edge set
+    struct pair_hash {
+      inline std::size_t operator()(const std::pair<gno_t,gno_t>& v) const {
+        return v.first * 10 + v.second;
```

Comment:
Why this choice of hash function?

Reply:
I didn't spend too much time on this hash function. I think my reasoning behind it was that an edge from A to B should be unlikely to get the same hash as an edge from B to A. Although, I do insert the edges so that they are in sorted order, so that case should never occur. I am very much open to changing this.